### DISCUSSION TITLE
Nested Menus

### DISCUSSION THREAD

**Discussion by Jinnified** [mantinedev/mantine#2307] [Feature requests]
hi folks, Mantine is awesome and it's just a breeze to use, loving it so far

just wondering, if you guys ever plan to make the Menu component nestable, so that multi-level menu nav can be a thing?

cheers

**Comment by rtivital** [comment]
Hi @Jinnified, thanks for kind words! Nested menus are in our backlog, I cannot say though when they are going to be released.

**Comment by atumas-bananamilk** [comment]
Really needed component!

**Comment by davigmacode** [comment]
It currently works with `closeOnItemClick={false}`, but still has odd behavior

![Nested Menu](https://bafybeiewztxd3tmvmzxsqoy7moo66kqqengnk5tzzrnewjq7eqvzl4hlv4.ipfs.nftstorage.link/)

```
<Menu width={200} shadow="md" position="bottom-start" closeOnItemClick={false}>
  <Menu.Target>
    <Button>Toggle menu</Button>
  </Menu.Target>

  <Menu.Dropdown>
    <Menu.Item>Menu Level 1</Menu.Item>

    <Menu width={200} shadow="md" position="right-start" closeOnItemClick={false}>
      <Menu.Target>
        <Menu.Item rightSection={<IconChevronRight size={18}/>}>
          Menu Level 1
        </Menu.Item>
      </Menu.Target>

      <Menu.Dropdown>
        <Menu.Item>Menu Level 2</Menu.Item>
        <Menu.Item>Menu Level 2</Menu.Item>
        <Menu width={200} shadow="md" position="right-start" closeOnItemClick={false}>
          <Menu.Target>
            <Menu.Item rightSection={<IconChevronRight size={18}/>}>
              Menu Level 2
            </Menu.Item>
          </Menu.Target>

          <Menu.Dropdown>
            <Menu.Item>Menu Level 3</Menu.Item>
            <Menu.Item>Menu Level 3</Menu.Item>
          </Menu.Dropdown>
        </Menu>
      </Menu.Dropdown>
    </Menu>
  </Menu.Dropdown>
</Menu>
```

**Reply by nwparker** [reply]
> It currently works with `closeOnItemClick={false}`, but still has odd behavior
> 
> ![Nested Menu](https://camo.githubusercontent.com/69f7dec3545d5bfe8dd4627a1b6a5746bd3be56aa76cce679f9105a12ff9f17b/68747470733a2f2f6261667962656965777a74786433746d766d7a7873716f79376d6f6f36366b7171656e676e6b35747a7a726e65776a71376571767a6c34686c76342e697066732e6e667473746f726167652e6c696e6b2f)
> 
> ```
> <Menu width={200} shadow="md" position="bottom-start" closeOnItemClick={false}>
>   <Menu.Target>
>     <Button>Toggle menu</Button>
>   </Menu.Target>
> 
>   <Menu.Dropdown>
>     <Menu.Item>Menu Level 1</Menu.Item>
> 
>     <Menu width={200} shadow="md" position="right-start" closeOnItemClick={false}>
>       <Menu.Target>
>         <Menu.Item rightSection={<IconChevronRight size={18}/>}>
>           Menu Level 1
>         </Menu.Item>
>       </Menu.Target>
> 
>       <Menu.Dropdown>
>         <Menu.Item>Menu Level 2</Menu.Item>
>         <Menu.Item>Menu Level 2</Menu.Item>
>         <Menu width={200} shadow="md" position="right-start" closeOnItemClick={false}>
>           <Menu.Target>
>             <Menu.Item rightSection={<IconChevronRight size={18}/>}>
>               Menu Level 2
>             </Menu.Item>
>           </Menu.Target>
> 
>           <Menu.Dropdown>
>             <Menu.Item>Menu Level 3</Menu.Item>
>             <Menu.Item>Menu Level 3</Menu.Item>
>           </Menu.Dropdown>
>         </Menu>
>       </Menu.Dropdown>
>     </Menu>
>   </Menu.Dropdown>
> </Menu>
> ```

Here's a small fix for that odd behavior that works with v7.11. Use this instead of `Menu.Item`
```
const MenuItem = Menu.Item.withProps({
  mod: { hovered: false },
  onMouseOver: (e) => {
    e.currentTarget.setAttribute('data-hovered', 'true');
  }
});
```

**Comment by waltershewmake** [comment]
I would love to see this feature in an upcoming release

**Comment by farhan-hafizh** [comment]
Hi, is there any update?

**Comment by vulebaolong** [comment]
You can refer to your code as follows

important: `withinPortal`
https://mantine.dev/core/popover/#nested-popovers

Nested popovers require children rendering without Portal. Usually, you should disable portal with props of the component that renders popover content, for example, Select has comboboxProps={{ withinPortal: false }} prop. Check documentation of the component that you are using to render popover content to find out how to disable the portal. If the portal is not disabled, outside click will close all popovers.

Example:

```js
{/* parent */}
<Popover width={300} position="bottom" withArrow shadow="md">
  <Popover.Target>
    <Button>Toggle popover</Button>
  </Popover.Target>
  <Popover.Dropdown bg="var(--mantine-color-body)">

   {/* children */}
    <Popover width={300} position="bottom" withArrow shadow="md" withinPortal={false}>
      <Popover.Target>
        <Button>Toggle popover</Button>
      </Popover.Target>
      <Popover.Dropdown bg="var(--mantine-color-body)">
        <Select
          label="Select within Popover"
          placeholder="Select within Popover"
          comboboxProps={{ withinPortal: false }}
          data={["React", "Angular", "Svelte", "Vue"]}
        />
        <Select
          label="Select within Popover"
          placeholder="Select within Popover"
          comboboxProps={{ withinPortal: false }}
          data={["React", "Angular", "Svelte", "Vue"]}
        />
        <Select
          label="Select within Popover"
          placeholder="Select within Popover"
          comboboxProps={{ withinPortal: false }}
          data={["React", "Angular", "Svelte", "Vue"]}
        />
      </Popover.Dropdown>
    </Popover>

    <Select
      label="Select within Popover"
      placeholder="Select within Popover"
      comboboxProps={{ withinPortal: false }}
      data={["React", "Angular", "Svelte", "Vue"]}
    />
    <Select
      label="Select within Popover"
      placeholder="Select within Popover"
      comboboxProps={{ withinPortal: false }}
      data={["React", "Angular", "Svelte", "Vue"]}
    />
    <Select
      label="Select within Popover"
      placeholder="Select within Popover"
      comboboxProps={{ withinPortal: false }}
      data={["React", "Angular", "Svelte", "Vue"]}
    />
  </Popover.Dropdown>
</Popover>
```

**Comment by clicktodev** [comment]
@rtivital can we please get this soon. This is a blocker for migrating one of ours apps to Mantine because we use nested menus a lot.

we're currently using the [radix implementation](https://www.radix-ui.com/primitives/docs/components/dropdown-menu)
![image](https://github.com/user-attachments/assets/204f6a08-25c1-48c2-8243-dab8c6580355)

**Comment by chinnuu05** [comment]
Any updates on this? I would really like to see this in Mantine, but I understand how it is with open source projects

**Reply by rtivital** [reply]
https://alpha.mantine.dev/changelog/8-0-0/#menu-with-submenus

**Reply by clicktodev** [reply]
@rtivital does it support infinite nesting or is it only one level?

**Reply by rtivital** [reply]
Infinite.

**Reply by clicktodev** [reply]
@rtivital you might wanna take a look at this https://github.com/radix-ui/primitives/issues/1749 discussion to improve the behavior on narrow screen widths 

bug:

![image](https://github.com/user-attachments/assets/b7a3bb5a-fc9f-4952-8a85-9d427a826913)

**Reply by engin-can** [reply]
Also, the chevrons aren't pointing the right way when you set `position="left-start"`

<img width="353" height="316" alt="CleanShot 2025-08-19 at 18 20 27" src="https://github.com/user-attachments/assets/1a6c8633-b7c5-4827-a928-14fdb32d2739" />

**Reply by clicktodev** [reply]
that's the correct behavior
